designs/ami-selector-reqs.md

```diff
@@ -0,0 +1,244 @@
+
+```yaml
+amiSelectorTerms:
+  - selector:
+      ssmParameter: "/my-org/amis/custom-ml-drivers"
+    requirements:
+      - key: kubernetes.io/arch
+        operator: In
+        values: [ 'amd64' ]
+      - key: "karpenter.k8s.aws/instance-gpu-count"
+        operator: "Exists"
+```
+
+* **Pro**: The association between an AMI and its requirements is declarative and self-contained.
+* **Pro**: Clear separation of concerns between AMI selection and scheduling logic.
+* **Con**: Breaking change to the `EC2NodeClass` schema. Migration could be handled via a new API version or an opt-in
+  feature gate.
+
+### SSM-Encoded Requirements
+
+This solution suggests embedding scheduling requirements directly into the SSM Parameter Store metadata associated with
+an AMI.
+
+Given an SSM parameter (e.g. `/my-org/amis/custom-ml-drivers`)
+
+One could then encode scheduling requirements into the SSM parameter as a JSON string, for example:
+
+```
+{"id":"ami-02211404f5aa0b76a","requirements":[{"key":"kubernetes.io/arch","operator":"In","values":["arm64"]},{"key":"karpenter.k8s.aws/instance-gpu-count","operator":"DoesNotExist"},{"key":"karpenter.k8s.aws/instance-accelerator-count","operator":"DoesNotExist"}]}
+```
+
+<details>
+<summary>Click to see formatted JSON...</summary>
+<pre><code>
+{
+    "id": "ami-02211404f5aa0b76a",
+    "requirements": [
+        {
+            "key": "kubernetes.io/arch",
+            "operator": "In",
+            "values": [
+                "arm64"
+            ]
+        },
+        {
+            "key": "karpenter.k8s.aws/instance-gpu-count",
+            "operator": "DoesNotExist"
+        },
+        {
+            "key": "karpenter.k8s.aws/instance-accelerator-count",
+            "operator": "DoesNotExist"
+        }
+    ]
+}
+</code></pre>
+</details>
+
+Karpenter would then resolve the AMI ID and the associated requirements from the SSM parameter, passing the discovered
+requirements on to the resolver, similar to how it does for alias.
+
+* **Pro**: Consistent with the existing `alias` behavior.
+* **Neutral**: Introduces new conventions around SSM parameter format when used as an `amiSelectorTerm`.
+  SSM parameters for well known AMI families (e.g. `bottlerocket`) do not follow this proposed convention.
+* **Con**: SSM Parameters have a maximum character limit of 2048, which imposes a hard cap on the number and
+  complexity of requirements that can be encoded. While this limit is generous and unlikely to be exceeded in typical
+  use cases, it may constrain more complex configurations or future extensibility.
+* **Con**: Reduced transparency. Requirements are no longer visible in the `EC2NodeClass` manifest, until after
+  the `status` has been reconciled.
+* **Con**: Not extensible to other `amiSelectorTerms`.
+
+### A Combination of 1 & 2
+
+Support `requirements` in the current schema (Option 1), and plan for a future breaking change to adopt a nested
```

Comment:
I'm actually not convinced we need to plan for a breaking change - what you laid out in option 1 convinces me at least that it makes sense as a long term solution.